sdclientapi/__init__.py

```diff
@@ -17,33 +18,44 @@
     Submission,
 )
 
-proxyvmname = "sd-proxy"
+DEFAULT_PROXY_VM_NAME = "sd-proxy"
+DEFAULT_REQUEST_TIMEOUT = 20  # 20 seconds
+DEFAULT_DOWNLOAD_TIMEOUT = 60 * 60  # 60 minutes
 
 
-def json_query(data: str) -> str:
+class RequestTimeoutError(Exception):
+    """
+    Error raisted if a request times out.
+    """
+
+    def __init__(self) -> None:
+        super().__init__("The request timed out.")
+
+
+def json_query(proxy_vm_name: str, data: str, timeout: Optional[int] = None) -> str:
     """
     Takes a json based query and passes to the network proxy.
     Returns the JSON output from the proxy.
     """
-    global proxyvmname
-    config = configparser.ConfigParser()
-    try:
-        if os.path.exists("/etc/sd-sdk.conf"):
-            config.read("/etc/sd-sdk.conf")
-            proxyvmname = config["proxy"]["name"]
-    except Exception:
-        pass  # We already have a default name
-
     p = Popen(
-        ["/usr/lib/qubes/qrexec-client-vm", proxyvmname, "securedrop.Proxy"],
+        ["/usr/lib/qubes/qrexec-client-vm", proxy_vm_name, "securedrop.Proxy"],
         stdin=PIPE,
         stdout=PIPE,
         stderr=PIPE,
     )
     p.stdin.write(data.encode("utf-8"))
-    stdout, _ = p.communicate()  # type: (bytes, bytes)
-    output = stdout.decode("utf-8")
-    return output.strip()
+
+    try:
+        stdout, _ = p.communicate(timeout=timeout)  # type: (bytes, bytes)
+    except TimeoutExpired:
+        try:
+            p.terminate()
+        except Exception:
+            pass
+        raise RequestTimeoutError
```

Comment:
we should add additional tests to exercise this path, i.e. verifying that the timeout exception is raised when a request takes longer than `DEFAULT_REQUEST_TIMEOUT`

Reply:
fixed